controllers/cluster_controller_postapply.go

```diff
@@ -0,0 +1,199 @@
+		}
+		err = ApplyYAMLWithNamespace(ctx, c, data, "")
+		if err != nil {
+			logger.Error(err, "Failed applying PostApply secret to cluster")
+			continue
+		}
+
+		logger.Info("Successfully applied post-apply addon", "Secret", addon.Namespace+"/"+addon.Name)
+
+		cluster.GetAnnotations()
+
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+		annotations[PostApplyAnnotation] = time.Now().Format(time.RFC3339)
+	}
+	return nil
+}
+
+// ApplyYAMLWithNamespace applies the provided YAML as unstructured data with the given client.
+// The data may be a single YAML document or multidoc YAML. This function is idempotent.
+// When a non-empty namespace is provided then all objects are assigned the namespace prior to being created.
+func ApplyYAMLWithNamespace(ctx context.Context, c client.Client, data []byte, namespace string) error {
+	return ForEachObjectInYAML(ctx, c, data, namespace, func(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
+		// Create the object on the API server.
+		if err := c.Create(ctx, obj); err != nil {
+			// The create call is idempotent, so if the object already exists
+			// then do not consider it to be an error.
+			if !apierrors.IsAlreadyExists(err) {
+				return errors.Wrapf(
+					err,
+					"failed to create object %s %s/%s",
+					obj.GroupVersionKind(),
+					obj.GetNamespace(),
+					obj.GetName())
+			}
+		}
+		return nil
+	})
+}
+
+// ForEachObjectInYAMLActionFunc is a function that is executed against each
+// object found in a YAML document.
+// When a non-empty namespace is provided then the object is assigned the
+// namespace prior to any other actions being performed with or to the object.
+type ForEachObjectInYAMLActionFunc func(context.Context, client.Client, *unstructured.Unstructured) error
+
+// ForEachObjectInYAML excutes actionFn for each object in the provided YAML.
+// If an error is returned then no further objects are processed.
+// The data may be a single YAML document or multidoc YAML.
+// When a non-empty namespace is provided then all objects are assigned the
+// the namespace prior to any other actions being performed with or to the
+// object.
+func ForEachObjectInYAML(
+	ctx context.Context,
+	c client.Client,
+	data []byte,
+	namespace string,
+	actionFn ForEachObjectInYAMLActionFunc) error {
+
+	chanObj, chanErr := DecodeYAML(data)
+	for {
+		select {
+		case obj := <-chanObj:
+			if obj == nil {
+				return nil
+			}
+			if namespace != "" {
+				obj.SetNamespace(namespace)
+			}
+			if err := actionFn(ctx, c, obj); err != nil {
+				return err
+			}
+		case err := <-chanErr:
+			if err == nil {
+				return nil
+			}
+			return errors.Wrap(err, "received error while decoding yaml to delete from server")
+		}
+	}
+}
+
+// DecodeYAML unmarshals a YAML document or multidoc YAML as unstructured
+// objects, placing each decoded object into a channel.
+func DecodeYAML(data []byte) (<-chan *unstructured.Unstructured, <-chan error) {
```

Comment:
This function should be moved to the `util/yaml` package as well.

I also don't know if we need to use channels here given that (unless I missed it) the work we do is sequential.
The function can just return a slice of unstructured objects and an `error`.

wdyt?